File: task4/GO/main.go
package main

import (
	"fmt"
)

// описывает группу избирателей с одинаковыми предпочтениями
type kGroup struct {
	Count       int
	Preferences []string
}

// метод Борда
func mBorda(n int, groups []kGroup, allCandid []string) {
	scores := make(map[string]int)

	// Инициализация баллов
	for _, name := range allCandid {
		scores[name] = 0
	}

	// Подсчет баллов Борда
	for _, group := range groups {
		for i := 0; i < n; i++ {
			points := n - 1 - i
			scores[group.Preferences[i]] += points * group.Count
		}
	}

	winner := ""
	maxPoints := -1

	fmt.Println("Результаты метода Борда")

	for name, score := range scores {
		fmt.Printf("%s: %d очков\n", name, score)
		if score > maxPoints {
			maxPoints = score
			winner = name
		}
	}
	fmt.Printf("Победитель по Борда: %s\n", winner)
}

// mCondorcet реализует метод Кондорсе
func mCondorcet(candidates []string, groups []kGroup) {
	winner := "не определен (парадокс Кондорсе)"

	for _, candA := range candidates {
		isWin := true

		for _, candB := range candidates {
			if candA == candB {
				continue
			}

			votesA := 0
			votesB := 0

			for _, group := range groups {
				posA, posB := -1, -1

				// поиск позиций
				for index, pref := range group.Preferences {
					if pref == candA {
						posA = index
					}
					if pref == candB {
						posB = index
					}
				}

				if posA < posB {
					votesA += group.Count
				} else {
					votesB += group.Count
				}
			}

			if votesA <= votesB {
				isWin = false
				break
			}
		}

		if isWin {
			winner = candA
			break
		}
	}

	fmt.Println("Результаты метода Кондорсе")
	fmt.Printf("Победитель по Кондорсе: %s\n", winner)
}

func main() {
	var n, k int

	fmt.Print("Введите количество кандидатов (n): ")
	fmt.Scan(&n)

	candidates := make([]string, n)
	fmt.Print("Введите имена кандидатов через пробел: ")
	for i := 0; i < n; i++ {
		fmt.Scan(&candidates[i])
	}

	fmt.Print("Введите количество групп избирателей (k): ")
	fmt.Scan(&k)

	groups := make([]kGroup, k)
	for i := 0; i < k; i++ {
		fmt.Printf("Группа %d. Количество избирателей: ", i+1)
		fmt.Scan(&groups[i].Count)

		fmt.Print("Ранжирование (через пробел): ")
		groups[i].Preferences = make([]string, n)
		for j := 0; j < n; j++ {
			fmt.Scan(&groups[i].Preferences[j])
		}
	}

	mBorda(n, groups, candidates)
	mCondorcet(candidates, groups)
}

File: task4/CPP/main.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>

using namespace std;

// данные об избирателях
struct kGroup {
    int count;
    vector<string> preferences;
};

// Метод Борда
void mBorda(int n, const vector<kGroup>& groups, const vector<string>& allCandid) {
    map<string, int> scores;

    // Иницилизация баллов
    for (const auto& name : allCandid) {
        scores[name] = 0;
    }

    // Подсчет баллов Борда
    for (const auto& group : groups) {
        for (int i = 0; i < n; ++i) {
            int points = n - 1 - i;
            scores[group.preferences[i]] += points * group.count;
        }
    }

    string winner;
    int maxPoints = -1;

    cout << "Результатф метода Борда" << endl;
    for (const auto& [name, score] : scores) {
        cout << name << ": " << score << " очков" << endl;
        if (score > maxPoints) {
            maxPoints = score;
            winner = name;
        }
    }
    cout << "Победитель по Борда: " << winner << endl;
}

// Метод Кондорсе
void mCondorcet(const vector<string>& candidates, const vector<kGroup>& groups) {
    string winner = "не определен (парадокс Кондорсе)";

    for (const auto& candA : candidates) {
        bool isWin = true;

        for (const auto& candB : candidates) {
            if (candA == candB) continue;

            int votesA = 0;
            int votesB = 0;

            for (const auto& group : groups) {
                int posA = -1, posB = -1;

                // поиск позиций
                int index = 0;
                for (const auto& pref : group.preferences) {
                    if (pref == candA) posA = index;
                    if (pref == candB) posB = index;
                    ++index;
                }

                if (posA < posB) votesA += group.count;
                else votesB += group.count;
            }

            if (votesA <= votesB) {
                isWin = false;
                break;
            }
        }

        if (isWin) {
            winner = candA;
            break;
        }
    }

    cout << "Результаты метода Кондорсе" << endl;
    cout << "Победитель по Кондорсе: " << winner  << endl;
}

int main() {
    int n, k;
    cout << "Введите количество кандидатов (n): ";
    cin >> n;

    vector<string> candidates(n);
    cout << "Введите имена кандидатов через пробел: ";
    for (auto& name : candidates) {
        cin >> name;
    }

    cout << "Введите количество групп избирателей (k): ";
    cin >> k;

    vector<kGroup> groups(k);
    for (int i = 0; i < k; ++i) {
        cout << "Группа " << i + 1 << ". Количество избирателей: ";
        cin >> groups[i].count;
        
        cout << "Ранжирование (через пробел): ";
        groups[i].preferences.resize(n);
        for (auto& pref : groups[i].preferences) {
            cin >> pref;
        }
    }

    mBorda(n, groups, candidates);
    mCondorcet(candidates, groups);

    return 0;
}